src/services/projects.js

```diff
@@ -64,10 +86,42 @@ angular.module('openshiftCommonServices')
                               });
                     });
           },
+
+          // List the projects the user has access to. This method returns
+          // cached data if the projects had previously been fetched to avoid
+          // requesting them again and again, which is a problem for admins who
+          // might have hundreds or more.
+          list: function(forceRefresh) {
+            if (cachedProjectData && !forceRefresh) {
+              Logger.debug("ProjectsService: returning cached project data");
+              return $q.when(cachedProjectData);
+            }
+
+            Logger.debug("ProjectsService: listing projects, force refresh", forceRefresh);
+            return DataService.list('projects', {}).then(function(projectData) {
+              cachedProjectData = projectData;
+              return projectData;
+            }, function(error) {
+              // If the request fails, don't try to list projects again without `forceRefresh`.
+              cachedProjectData = {};
+            });
+          },
+
+          watch: function(context, callback) {
+            // Wrap `DataService.watch` so we can update the cached projects
+            // list on changes. TODO: We might want to disable watches entirely
+            // if we know the project list is large.
+            return DataService.watch("projects", context, function(projectData) {
+              cachedProjectData = projectData;
+              callback(projectData);
```

Comment:
A call to watch could return the current cachedProjectData. And invoke the callback again when projects change. This would be a bit more complex as this would need to handle unwatch as well.